File: src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionFlat.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.JsonElementStreamer;
import co.nlighten.jsontransform.adapters.JsonAdapter;
import co.nlighten.jsontransform.functions.common.ArgType;
import co.nlighten.jsontransform.functions.common.FunctionContext;
import co.nlighten.jsontransform.functions.common.TransformerFunction;
import co.nlighten.jsontransform.functions.annotations.Aliases;
import co.nlighten.jsontransform.functions.annotations.Documentation;
import co.nlighten.jsontransform.functions.annotations.InputType;
import co.nlighten.jsontransform.functions.annotations.OutputType;

import java.util.Objects;
import java.util.stream.Stream;

/*
 * For tests
 * @see TransformerFunctionFlatTest
 */
@Aliases("flat")
@Documentation("Flatten an array of arrays (non array elements will remain, all `null` elements are removed from result)")
@InputType(ArgType.Array)
@OutputType(ArgType.Array)
public class TransformerFunctionFlat<JE, JA extends Iterable<JE>, JO extends JE> extends TransformerFunction<JE, JA, JO> {
    public TransformerFunctionFlat(JsonAdapter<JE, JA, JO> adapter) {
        super(adapter);
    }
    @Override
    public Object apply(FunctionContext<JE, JA, JO> context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null) return null;

        return JsonElementStreamer.fromTransformedStream(context, streamer.stream()
            .flatMap(itm -> {
                if (adapter.isNull(itm)) {
                    return Stream.empty();
                } else if (ARRAY.is(itm)) {
                    return ARRAY.stream((JA)itm);
                }
                return Stream.of(itm);
            })
            .filter(Objects::nonNull));
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionFlat.ts
import { asAsyncSequence, asyncSequenceOf, emptyAsyncSequence } from "@wortise/sequency";
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import { FunctionDescription } from "./common/FunctionDescription";
import FunctionContext from "./common/FunctionContext";
import { isNullOrUndefined } from "../JsonHelpers";
import JsonElementStreamer from "../JsonElementStreamer";

const DESCRIPTION: FunctionDescription = {
  aliases: ["flat"],
  inputType: ArgType.Array,
  description: "",
  outputType: ArgType.Array,
};
class TransformerFunctionFlat extends TransformerFunction {
  constructor() {
    super(DESCRIPTION);
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null) return null;

    return JsonElementStreamer.fromTransformedStream(
      context,
      streamer
        .stream()
        .flatMap(itm => {
          if (isNullOrUndefined(itm)) {
            return emptyAsyncSequence();
          } else if (Array.isArray(itm)) {
            return asAsyncSequence(itm);
          }
          return asyncSequenceOf(itm);
        })
        .filter(el => !isNullOrUndefined(el)),
    );
  }
}

export default TransformerFunctionFlat;
